app/models/trade_import.rb

```diff
@@ -41,12 +51,63 @@ def dry_run
 
   def csv_template
     template = <<-CSV
-      date*,ticker*,qty*,price*,currency,account,name
-      05/15/2024,AAPL,10,150.00,USD,Trading Account,Apple Inc. Purchase
-      05/16/2024,GOOGL,-5,2500.00,USD,Investment Account,Alphabet Inc. Sale
-      05/17/2024,TSLA,2,700.50,USD,Retirement Account,Tesla Inc. Purchase
+      date*,ticker*,exchange_operating_mic,currency,qty*,price*,account,name
+      05/15/2024,AAPL,XNAS,USD,10,150.00,Trading Account,Apple Inc. Purchase
+      05/16/2024,GOOGL,XNAS,USD,-5,2500.00,Investment Account,Alphabet Inc. Sale
+      05/17/2024,TSLA,XNAS,USD,2,700.50,Retirement Account,Tesla Inc. Purchase
     CSV
 
     CSV.parse(template, headers: true)
   end
+
+  private
+    def find_or_create_security(ticker:, exchange_operating_mic:)
+      # Normalize empty string to nil for consistency
+      exchange_operating_mic = nil if exchange_operating_mic.blank?
+
+      # First try to find an exact match in our DB
+      internal_security = Security.find_by(ticker:, exchange_operating_mic:)
+      return internal_security if internal_security.present?
+
+      # If no exact match and no exchange_operating_mic was provided, try to find any security with the same ticker
+      if exchange_operating_mic.nil?
+        internal_security = Security.where(ticker:).first
+        if internal_security.present?
+          internal_security.update!(exchange_operating_mic: nil)
+          return internal_security
+        end
+      end
+
+      # If we couldn't find the security locally and the provider isn't available, create with provided info
+      return Security.create!(ticker: ticker, exchange_operating_mic: exchange_operating_mic) unless Security.security_prices_provider.present?
+
+      # Cache provider responses so that when we're looping through rows and importing, we only hit our provider for the unique combinations of ticker / exchange_operating_mic
+      cache_key = [ ticker, exchange_operating_mic ]
+
+      @provider_securities_cache ||= {}
+
+      provider_security = @provider_securities_cache[cache_key] ||= begin
+        response = Security.security_prices_provider.search_securities(
+          query: ticker,
+          exchange_operating_mic: exchange_operating_mic
+        )
+
+        return nil unless response.success?
+
+        response.securities.first
+      end
+
+      return Security.create!(ticker: ticker, exchange_operating_mic: exchange_operating_mic) if provider_security.nil?
```

Comment:
We need to explicitly set `exchange_operating_mic: nil` here as I had in my snippet below to prevent invalid exchange mics from being persisted to the DB.